process_basic.go

```diff
@@ -0,0 +1,208 @@
+package jasper
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+
+	"github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+)
+
+type basicProcess struct {
+	info          ProcessInfo
+	cmd           *exec.Cmd
+	err           error
+	id            string
+	hostname      string
+	opts          CreateOptions
+	tags          map[string]struct{}
+	triggers      ProcessTriggerSequence
+	waitProcessed chan struct{}
+	initialized   chan struct{}
+	sync.RWMutex
+}
+
+func newBasicProcess(ctx context.Context, opts *CreateOptions) (Process, error) {
+	id := uuid.Must(uuid.NewV4()).String()
+	opts.AddEnvVar(EnvironID, id)
+
+	cmd, err := opts.Resolve(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem building command from options")
+	}
+
+	p := &basicProcess{
+		id:            id,
+		opts:          *opts,
+		cmd:           cmd,
+		tags:          make(map[string]struct{}),
+		waitProcessed: make(chan struct{}),
+		initialized:   make(chan struct{}),
+		triggers:      ProcessTriggerSequence{},
+	}
+	p.hostname, _ = os.Hostname()
+
+	for _, t := range opts.Tags {
+		p.Tag(t)
+	}
+
+	p.RegisterTrigger(ctx, makeOptionsCloseTrigger())
+
+	err = cmd.Start()
+	if err != nil {
+		return nil, errors.Wrap(err, "problem creating command")
+	}
+
+	p.info.ID = p.id
+	p.info.Options = p.opts
+	p.info.Host = p.hostname
+
+	go p.transition(ctx, cmd)
+
+	opts.started = true
+
+	return p, nil
+}
+
+func (p *basicProcess) transition(ctx context.Context, cmd *exec.Cmd) {
+	waitFinished := make(chan error)
+
+	initialize := func() {
```

Comment:
I feel like you only call initialize once, so you might as well move the function call/body there so it  reads more realistically.

Reply:
Moved the definitions of `initialize()` and `finish()` close to their call-site, but kept the variables instead of using something like `func() { ... }()`. I think this serves as good self-documentation about what the two blocks of code do in terms of transitioning the process through phases.